android/src/main/kotlin/im/nfc/flutter_nfc_kit/FlutterNfcKitPlugin.kt

```diff
@@ -284,10 +293,276 @@ class FlutterNfcKitPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
+    }
+
+    private fun readSector(result: Result, sectorIndex: Int, authenticateKeyA: ByteArray?) {
+        if (isMifareClassic) {
+            val mifareClassic = MifareClassic.get(tagTechnology?.tag)
+            try {
+                val sectorAuthenticateKeyA: ByteArray? = if (authenticateKeyA?.isEmpty() == null) {
+                    MifareClassic.KEY_DEFAULT
+                } else {
+                    authenticateKeyA
+                }
+                mifareClassic.connect()
+                mifareClassic.authenticateSectorWithKeyA(sectorIndex,sectorAuthenticateKeyA)
+                val sector = ByteUtils.mifareClassPrintEntireBlock(mifareClassic, sectorIndex)
+                activity?.runOnUiThread { result.success(sector) }
+            } catch (e: Exception) {
+                Log.e(TAG, "read sector error", e)
+                activity?.runOnUiThread { result.error("502", e.localizedMessage, null) }
+            } finally {
+                mifareClassic.close()
+            }
+        }
+    }
+
+    private fun readAll(result: Result, authenticateKeyA: ByteArray?) {
+        val response = mutableMapOf<Int, List<String>>()
+        if (isMifareClassic) {
+            val mifareClassic = MifareClassic.get(tagTechnology?.tag)
+            try {
+                val sectorAuthenticateKeyA: ByteArray? = if (authenticateKeyA?.isEmpty() == null) {
+                    MifareClassic.KEY_DEFAULT
+                } else {
+                    authenticateKeyA
+                }
+                mifareClassic.connect()
+                for (i in 0 until mifareClassic.sectorCount) {
+                    mifareClassic.authenticateSectorWithKeyA(i, sectorAuthenticateKeyA)
+                    response[i] = ByteUtils.mifareClassPrintEntireBlock(mifareClassic, i)
+                }
+                activity?.runOnUiThread { result.success(response) }
+            } catch (e:java.lang.Exception) {
+                Log.e(TAG, "Read MifareClassic All Error: ", e)
+                activity?.runOnUiThread { result.error("503", e.localizedMessage, null) }
+            } finally {
+                mifareClassic.close()
+            }
+        } else if (isMifareUltralight) {
+            val  mifareUltralight = MifareUltralight.get(tagTechnology?.tag)
+            try {
+                mifareUltralight.connect()
+                for (i in 0 until 16) {
+                    response[i] = ByteUtils.mifareUltralightPrintEntireBlock(mifareUltralight, i)
+                }
+                activity?.runOnUiThread { result.success(response) }
+            } catch (ex: Exception) {
+                Log.e(TAG, "Read MifareUltralight All Error: ", ex)
+                activity?.runOnUiThread { result.error("503", ex.localizedMessage, null) }
+            } finally {
+                mifareUltralight.close()
+            }
+        }
+    }
+    private fun writeBlock(result: Result, blockIndex: Int, authenticateKeyA: ByteArray?, message: ByteArray) {
```

Comment:
What about key B?

Reply:
Key B is necessary? I see the doc about MifareClassic `authenticateKeyB` as optional.